### ISSUE TITLE
Sanity check for assertion instanceOf

### ISSUE THREAD

**Issue by fgarcia** [chaijs/chai#893]
The current implementation for the `instanceOf` assert does not check if the given constructor is defined

https://github.com/chaijs/chai/blob/104a6b7db9e02d24cd296fbeddc3f55907e8fda0/lib/chai/core/assertions.js#L989

There is a common error I hit several times where an imported variable stops working. I use that variable to assert the instance of an object, like:

```javascript
import { Pathname } from 'core'
....
expect(x).to.be.instanceOf(Pathname)
```

When the import fails (silently) with an undefined value, the error thrown by the assertion shows a normal Javascript error

`TypeError: Cannot read property 'name' of undefined`

It would be great getting something more meaningful like

`ChaiError: instanceOf() needs a constructor. undefined value given`

**NOTE:**

I've realized there is no such thing as `ChaiError` error subtype. I am not sure how far the current code base goes to improve bad use cases

**Comment by vieiralucas** [comment]
Hello @fgarcia thanks for your issue.

IMHO this is a chai problem, I agree with you that this should provide a more meaningful error message.
Since we move the `getName` utility to it's own module ([chaijs/get-func-name](https://github.com/chaijs/get-func-name)), currently on `4.0.0-canary.1` this error would be throw from [here](https://github.com/chaijs/get-func-name/blob/b4a20e94373a86822e317ad510cd259592fce043/index.js#L33).

So, IMHO:
 -  `get-func-name` should throw its own exception if `undefined` is given. Maybe we should open an issue there
 - `instanceOf` should do something similar to [above](https://github.com/chaijs/chai/blob/104a6b7db9e02d24cd296fbeddc3f55907e8fda0/lib/chai/core/assertions.js#L707)

It should be very easy to do this, we just need to add a check [here](https://github.com/chaijs/chai/blob/104a6b7db9e02d24cd296fbeddc3f55907e8fda0/lib/chai/core/assertions.js#L990)

Am I missing something here? Do someone disagree?

**Comment by lucasfcosta** [comment]
Hi @fgarcia, thanks for your issue! 😄 

Well, I totally agree with @vieiralucas that we should fix this.

However, I think this is something that should be done within Chai's core.

IMO the whole point of having a separate module for `get-func-name` is making it modular enough to be used by anyone wanting to, including us. By throwing an error there we would be coupling what we want for `Chai` with the module's own purpose.

I think that it would be more adequate to just add a check to `getFunctionName` to avoid it throwing a `TypeError` and then just return `undefined` if `undefined` was passed to it. Then we could just add our own specific logic here at `Chai` to throw a custom [`AssertionError`](https://github.com/chaijs/assertion-error), since this is useful for us only, after all, not everyone using `get-func-name` will expect it to throw this kind of `Error`.

Feel free to share your thoughts :)

**Comment by vieiralucas** [comment]
I didnt mean to change get-func-name to throw a chai specific custom error.
I was just saying that it should not be exploding trying to access an property on undefined.
I guess returning undefined is good to, but as I said we should discuss it over there.
:smile:

**Comment by shvaikalesh** [comment]
Agreed that we should validate what is getting passed to `instanceOf`. However, with ES6's `@@hasInstance` that is not trivial task to do. Second operand of `instanceof` is allowed to be either an object with callable `@@hasInstance` or function (not necessary with `[[Construct]]`) with `prototype` object. First operand can be of any type, primitives included.

**Comment by meeber** [comment]
Fixed via #899
